### PR TITLE
Stop generating out full function type for registration, use decltype or infer it

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #33205 Bring up new-style registration API as wrapper around old-style
* **#33097 Stop generating out full function type for registration, use decltype or infer it**
* #33093 Delete unnecessary aliasAnalysis specification from operator registrations.

Previously, we had to specify full types because the functions we registering
might be overloaded, and the type was necessary to resolve the ambiguity.  I
disambiguate all of these names by mangling the names of the methods we
place on CPUType/CUDAType/TypeDefault with the overload name (these are
*internal* wrappers which are not user visible), and then can strip
the generation of full function types from the registration.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D19837898](https://our.internmc.facebook.com/intern/diff/D19837898)

### CODE REVIEW

**Comment by ezyang** [comment]
New registration looks like, e.g.,

```
  .op(torch::RegisterOperators::options()
    .schema("aten::im2col.out(Tensor self, int[2] kernel_size, int[2] dilation, int[2] padding, int[2] stride, *, Tensor(a!) out) -> Tensor(a!)")
    .impl_unboxedOnlyKernel<decltype(CPUType::im2col_out_out), &CPUType::im2col_out_out>(DispatchKey::CPUTensorId))
```

File: aten/src/ATen/common_with_cwrap.py [hunk 1]
```diff
@@ -30,8 +30,14 @@ def set_declaration_defaults(declaration):
         declaration['cname'] = declaration['name']
     if 'backends' not in declaration:
         declaration['backends'] = ['CPU', 'CUDA']
-    if 'api_name' not in declaration:
-        declaration['api_name'] = declaration['name']
+    assert 'api_name' not in declaration
+    declaration['api_name'] = declaration['name']
+    # NB: keep this in sync with gen_autograd.py
+    if declaration.get('overload_name'):
+        declaration['type_wrapper_name'] = "{}_{}".format(
```
Comment (gchanan) [inline aten/src/ATen/common_with_cwrap.py]: you can do this if you want, but for state like this that is totally derivable, I think it's clearer if you:
1) don't put the type_wrapper_name in the declaration, use a common function to compute it
2) don't use the environment, just pass the key explicitly to the CodeTemplate
(this is in the line with what @bhosmer did for gen_python_functions).

I do think this is okay given we haven't "cleaned up" function_wrapper yet, and this is in the existing style.

File: tools/autograd/gen_autograd.py [hunk 2]
```diff
@@ -111,6 +111,12 @@ def load_aten_declarations(path):
                                               for arg in declaration['arguments']]
         declaration['type_method_args'] = [arg['name'] for arg in declaration['arguments']]
         declaration['api_name'] = declaration['name']
+        # NB: keep this in sync with common_with_cwrap.py
+        if declaration.get('overload_name'):
```
Comment (gchanan) [inline tools/autograd/gen_autograd.py]: see, if you had a common function to compute this, you wouldn't need to worry about keeping this in sync :).

**Review by gchanan — APPROVED** [review]

File: tools/autograd/gen_autograd.py (hunk 2 above)
Comment (ezyang) [reply tools/autograd/gen_autograd.py]: Can I even import any of the crap from ATen python scripts though? I feel like the only thing we share right now is CodeTemplate...

Comment (gchanan) [reply tools/autograd/gen_autograd.py]: that's a good question I don't know the answer to -- it feels like a good thing to fix if it doesn't work, though!

Comment (ezyang) [reply tools/autograd/gen_autograd.py]: Looking at the imports, it seems that only CodeTemplate is shared, and you have to go through some shenanigans to get to it:

```
try:
    from src.ATen.code_template import CodeTemplate
except ImportError:
    from tools.shared.module_loader import import_module
    CodeTemplate = import_module('code_template', 'aten/src/ATen/code_template.py').CodeTemplate

```

Comment (bhosmer) [reply tools/autograd/gen_autograd.py]: @ezyang the shenanigans are in some places [but not others](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/gen.py#L19)... not sure why.

Comment (ezyang) [reply tools/autograd/gen_autograd.py]: files in ATen don't need these shenanigans as they are in the same network

**Review by smessmer — APPROVED** [review]
Looks good, with a slight preference for implementing @gchanan's proposal of not creating the new key. But it's also fine with me to land it in the current state.

**Pull request closed without merging**